sycl/unittests/pi/piInteropRetain.cpp

```diff
@@ -0,0 +1,69 @@
+//==--------------------- Wait.cpp --- queue unit tests --------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/opencl.h>
+#include <CL/sycl.hpp>
+#include <CL/sycl/backend/opencl.hpp>
+
+#include <detail/queue_impl.hpp>
+#include <gtest/gtest.h>
+#include <helpers/PiMock.hpp>
+
+namespace {
+using namespace cl::sycl;
+
+static bool QueueRetainCalled;
+pi_result unexpectedQueueRetain(pi_queue Queue) { return PI_INVALID_QUEUE; }
+pi_result expectedQueueRetain(pi_queue Queue) {
+  QueueRetainCalled = true;
+  return PI_SUCCESS;
+}
+
+bool preparePiMock(platform &Plt) {
+  if (Plt.is_host()) {
+    std::cout << "Not run on host - no PI events created in that case"
+              << std::endl;
+    return false;
+  }
+  if (detail::getSyclObjImpl(Plt)->getPlugin().getBackend() !=
+      backend::opencl) {
+    std::cout << "Not run on non-OpenCL backend" << std::endl;
+    return false;
+  }
+  return true;
+}
+
+TEST(PiInteropTest, CheckRetain) {
+  platform Plt{default_selector()};
+  if (!preparePiMock(Plt))
+    return;
+  context Ctx{Plt.get_devices()[0]};
+
+  unittest::PiMock Mock{Plt};
+
+  // The queue construction should not call to piQueueRetain. Instead
+  // piQueueCreate should return the "retained" queue.
+  QueueRetainCalled = false;
+  Mock.redefine<detail::PiApiKind::piQueueRetain>(unexpectedQueueRetain);
```

Comment:
```suggestion
  Mock.redefine<detail::PiApiKind::piQueueRetain>(expectedQueueRetain);
```
Shouldn't we pass `expectedQueueRetain` here? Since we check `QueueRetainCalled` later which cannot be changed in `unexpectedQueueRetain`.
I'm not sure this test needs `unexpectedQueueRetain` at all.

Reply:
The `unexpectedQueueRetain` would bail out with an error, so it is correct as is, but I can rewrite with just single "redefine" and counting.